### PR TITLE
[Text Translation] Add support for AAD authentication

### PR DESCRIPTION
**Pull request by MikeyMCZ**
# Description

Add support for AAD authentication for global Text Translation Endpoint which requires two special headers:
* `Ocp-Apim-ResourceId`
* `Ocp-Apim-Subscription-Region`

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 1]
```diff
@@ -122,7 +148,7 @@ class TextTranslationClient(ServiceClientGenerated):
          https://api.cognitive.microsofttranslator.com). Required.
     :type endpoint: str
     :param credential: Credential used to authenticate with the Translator service
-    :type credential: Union[AzureKeyCredential , TokenCredential , TranslatorCredential]
+    :type credential: Union[AzureKeyCredential , TokenCredential , TranslatorCredential, TranslatorAADCredential]
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: Can we instead support the TokenCredentials provided by the azure-identity package? If we do this, then you should be able to get all of that goodness and credential logic without having to maintain it in this library.

Comment (catalinaperalta) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: Also, if you now support AAD auth in your service, then you should specify that in the TypeSpec as well

Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: I am not sure what do you mean and what does it provides but I would guess it doesn't provide the two MT specific headers that we need when global MT endpoint is used. https://learn.microsoft.com/en-us/azure/ai-services/translator/reference/v3-0-reference#authentication-with-microsoft-entra-id

Comment (catalinaperalta) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: @xiangyan99 @pvaneck do we have support for this auth scenario in azure identity?

Comment (xiangyan99) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: @MikeyMCZ,  do you use same or different scopes/audiences for global endpoint vs custom endpoint?

Comment (xiangyan99) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: @catalinaperalta if I understand correctly, here we still want to use token credentials. The trick is in addition to the token itself, for global endpoint scenarios, we also need the information of `Resource ID` and `region`

Comment (xiangyan99) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: Btw, you may need to discuss with architects about 
1. If we need two different `TranslatorAADCredential` classes?
2. The field names of the credential class. To be clearer:
        self.tokenCredential = tokenCredential
        self.resourceId = resourceId
        self.region = region
3. Given these different combinations (endpoint + credentials), do we want to use overloads?

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 2]
```diff
@@ -56,6 +56,29 @@ def on_request(self, request: PipelineRequest) -> None:
...
 
+class TranslatorAADCredential:
+    """Credential for Translator Service when using AAD authentication."""
+
+    def __init__(self, tokenCredential: TokenCredential, resourceId: str, region: str) -> None:
+        self.tokenCredential = tokenCredential
+        self.resourceId = resourceId
+        self.region = region
+
+class TranslatorAADAuthenticationPolicy(BearerTokenCredentialPolicy):
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:67]: Can you add docs for `translatorCredential`

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 3]
```diff
@@ -56,6 +56,29 @@ def on_request(self, request: PipelineRequest) -> None:
         request.http_request.headers["Ocp-Apim-Subscription-Key"] = self.credential.key
         request.http_request.headers["Ocp-Apim-Subscription-Region"] = self.credential.region
 
+class TranslatorAADCredential:
+    """Credential for Translator Service when using AAD authentication."""
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:60]: Can you also add docs here for the properties?

File: sdk/translation/azure-ai-translation-text/tests/test_break_sentence.py [hunk 4]
```diff
@@ -21,7 +21,7 @@ def test_autodetect(self, **kwargs):
         response = client.find_sentence_boundaries(request_body=input_text_elements)
         assert response is not None
         assert response[0].detected_language.language == "en"
-        assert response[0].detected_language.score == 1
+        assert response[0].detected_language.score == 0.98
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/tests/test_break_sentence.py:24]: If this score isnt 100% reliable you could just check if it's >= 0.9

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py (hunk 1 above)
Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: _do you use same or different scopes/audiences for global endpoint vs custom endpoint?_
We are using the same scope for the token.

_if I understand correctly, here we still want to use token credentials. The trick is in addition to the token itself, for global endpoint scenarios, we also need the information of Resource ID and region_
That is correct. Same change was already accepted in dotnet and Java should be approved and merge soon.

File: sdk/translation/azure-ai-translation-text/tests/test_translation.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 # Licensed under the MIT License.
 # ------------------------------------
 
+import os
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/tests/test_translation.py:6]: Nit: I dont see where os was added?

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/aio/_patch.py
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/aio/_patch.py:1]: Same comment for the changes in this file, please document the properties

File: sdk/translation/azure-ai-translation-text/tests/test_translation.py (hunk 5 above)
Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/tests/test_translation.py:6]: Removed, it was added by one of my attempts for test

File: sdk/translation/azure-ai-translation-text/tests/test_break_sentence.py (hunk 4 above)
Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/tests/test_break_sentence.py:24]: Good idea, done.

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py (hunk 2 above)
Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:67]: Sure

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py (hunk 3 above)
Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:60]: Sure

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/aio/_patch.py
Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/aio/_patch.py:1]: done

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py (hunk 1 above)
Comment (MikeyMCZ) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:151]: _Given these different combinations (endpoint + credentials), do we want to use overloads?_
I am ok with any change that will be advised - I am not sure what is the best practice in the Python. Do you mind providing the code suggestion that I can accept?

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 6]
```diff
@@ -56,6 +56,36 @@ def on_request(self, request: PipelineRequest) -> None:
         request.http_request.headers["Ocp-Apim-Subscription-Key"] = self.credential.key
         request.http_request.headers["Ocp-Apim-Subscription-Region"] = self.credential.region
 
+class TranslatorAADCredential:
+    """Credential for Translator Service when using AAD authentication.
+    :param tokenCredential: An object which can provide an access token for the Translator Resource, such as a credential from
+        :mod:`azure.identity`
+    :type tokenCredential: ~azure.core.credentials.TokenCredential
+    :param resourceId: Azure Resource Id of the Translation Resource.
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:64]: We need to specify a type
```suggestion
    :param str resourceId: Azure Resource Id of the Translation Resource.
```

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 7]
```diff
@@ -56,6 +56,36 @@ def on_request(self, request: PipelineRequest) -> None:
...
+    :param region: Azure Region of the Translation Resource.
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:65]:
```suggestion
    :param str region: Azure Region of the Translation Resource.
```

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 8]
```diff
@@ -56,6 +56,36 @@ def on_request(self, request: PipelineRequest) -> None:
...
+        self.tokenCredential = tokenCredential
+        self.resourceId = resourceId
+        self.region = region
+
+class TranslatorAADAuthenticationPolicy(BearerTokenCredentialPolicy):
+    """Translator AAD Authentication Policy. Adds headers that are required by Translator Service
+    when global endpoint is used with AAD policy.
+    Ocp-Apim-Subscription-Region header contains region of the Translator resource.
+    Ocp-Apim-ResourceId header contains Azure resource Id - Translator resource.
+    :param credential: Translator AAD Credentials used to access Translator Resource for global Translator endpoint.
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:78]:
```suggestion
    :param credential: Translator AAD Credentials used to access Translator Resource for global Translator endpoint.
    :type tokenCredential: ~azure.core.credentials.TokenCredential
```

Comment (catalinaperalta) [reply sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:78]: Same applies to async

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 9]
```diff
@@ -56,6 +56,37 @@ def on_request(self, request: PipelineRequest) -> None:
...
+    :type tokenCredential: ~azure.core.credentials.TokenCredential
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:79]:
```suggestion
    :type credential: ~azure.core.credentials.TokenCredential
```

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/aio/_patch.py [hunk 10]
```diff
@@ -24,11 +25,46 @@ def patch_sdk():
...
+        self.resourceId = resourceId
+        self.region = region
+
+class AsyncTranslatorAADAuthenticationPolicy(AsyncBearerTokenCredentialPolicy):
+    """Translator AAD Authentication Policy. Adds headers that are required by Translator Service
+    when global endpoint is used with AAD policy.
+    Ocp-Apim-Subscription-Region header contains region of the Translator resource.
+    Ocp-Apim-ResourceId header contains Azure resource Id - Translator resource.
+    :param credential: Translator AAD Credentials used to access Translator Resource for global Translator endpoint.
+    :type tokenCredential: ~azure.core.credentials.TokenCredential
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/aio/_patch.py:48]:
```suggestion
    :type credential: ~azure.core.credentials.TokenCredential
```

**Review by catalinaperalta — APPROVED** [review]
LGTM. Please make sure to review this design with architects.

File: sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py [hunk 11]
```diff
@@ -56,6 +56,37 @@ def on_request(self, request: PipelineRequest) -> None:
...
+        self.tokenCredential = tokenCredential
+        self.resourceId = resourceId
+        self.region = region
+
+class TranslatorAADAuthenticationPolicy(BearerTokenCredentialPolicy):
+    """Translator AAD Authentication Policy. Adds headers that are required by Translator Service
+    when global endpoint is used with AAD policy.
+    Ocp-Apim-Subscription-Region header contains region of the Translator resource.
+    Ocp-Apim-ResourceId header contains Azure resource Id - Translator resource.
+    :param credential: Translator AAD Credentials used to access Translator Resource for global Translator endpoint.
```
Comment (catalinaperalta) [inline sdk/translation/azure-ai-translation-text/azure/ai/translation/text/_patch.py:78]:
```suggestion

    :param credential: Translator AAD Credentials used to access Translator Resource for global Translator endpoint.
```

**Pull request merged**